### ISSUE TITLE
[labs/ssr] opt-in de-dupe styles in declarative shadow DOM

### ISSUE THREAD

**Issue by AndrewJakubowicz** [lit/lit#4357]
### Should this be an RFC?

- [X] This is not a substantial change

### Which package is this a feature request for?

SSR (@lit-labs/ssr)

### Description

For reference, a platform solution: https://github.com/WICG/webcomponents/issues/939

Currently, each instance of a component requires all its styles to be inlined into the DSD. This means if you have the same component on the page multiple times, you repeat the styles per instance. This leads to duplicated styles being sent down the network.

Before the platform proposal exists, we should offer a solution.

### Proposed solution

This is not a full design, just some rough notes initially.
We can add an de-dupe styles option to Lit SSR.

SSR could inline a JavaScript utility, such as a custom element definition into the top of the page. E.g. `<style-module>`.
Then when emitting styles in DSD, wrap the first instance of a DSD style in `<style-module>`:

```
<style-module style-id="abc">
  <style>
    ...
   </style>
</style-module>
```

If these styles will be duplicated, any future emit of the styles contains just the `<style-module style-id="abc">` without any style contents.

This custom element fetches the styles from a data store, and adds them to the DSD.

### Considerations

This should be an option that can be turned on and off, because this solution requires JavaScript be run on the page – but the benefit is a reduced payload. For clients who run JavaScript, this would allow for a reduced network payload.

### Alternatives and Workarounds

Keep duplicating styles until browser platform adds support for this.

**Comment by bivens-dev** [comment]
Just spitballing a potentially JS free solution, could the same thing be achieved using link tags the components styles? 

Is it possible to reference a style tag on the page using ids and anchor links as a place extract common styles to.

**Comment by justinfagnani** [comment]
One way to do this is to add `<lit-adoptable-style>` and `<lit-adopt-styles>`tags. The former would register a stylesheet in a global cache, the later would read from the cache and add styles to its root node's `adoptedStyleSheets`.

There's an [issue open](https://github.com/webcomponents-cg/community-protocols/issues/72#issuecomment-2835940612) in WCCG for a community protocol around this, but I don't think it's necessary. We can just do this in Lit SSR and it will be compatible with any elements, since they won't have to know about either style-related custom element here, ie it'll work with any declarative shadow root, not just Lit-based elements.

Implementation note: In https://github.com/webcomponents-cg/community-protocols/issues/72#issuecomment-2835940612 I show an sketch element that's mean to go _after_ a `<style>` tag in order to work around the lack of a [childrenParsed callback](https://github.com/WICG/webcomponents/issues/809). That's ergonomically a bit weird, but this is meant for machine use, not human use. We could choose to spend a few more bytes and CPU cycles and wire up mutation observers and/or slotchange listeners to make the text content be inside the `<lit-adoptable-style>` element if we care.

cc @sorvell

**Comment by ChellappanRajan** [comment]
We enabled Deep Rendering using the lit-labs/nextjs package, but all components are inlining styles, which results in significantly higher initial page load times compared to client-side rendering with Web Components.

Is there any fix or workaround for this issue?

**Comment by sorvell** [comment]
[Here's another take on this](https://lit.dev/playground/#gist=da5c8db37fa08e748d60f2dc5510848c) which avoids using another style by just waiting for requestAnimationFrame, which is before paint

**Comment by justinfagnani** [comment]
@sorvell I think for various parser corner cases were going to want to use a real `<style>` tag, unless we escape and unescape the sheet text. `content` that contains HTML is the tricky bit, I think. 

But in order to apply sheets in the right order, we'll also probably want to got the style tag a non-standard type.

**Comment by ChellappanRajan** [comment]
> [Here's another take on this](https://lit.dev/playground/#gist=da5c8db37fa08e748d60f2dc5510848c) which avoids using another style by just waiting for requestAnimationFrame, which is before paint

Thanks for the suggestions, Since it's required JS will it cause FOUC?

**Comment by sorvell** [comment]
> Thanks for the suggestions, Since it's required JS will it cause FOUC?

No, the JS will be a small snippet inlined at the top of the page, executed synchronously and fast. This does not cause FOUC.

**Comment by ChellappanRajan** [comment]
> > Thanks for the suggestions, Since it's required JS will it cause FOUC?
> 
> No, the JS will be a small snippet inlined at the top of the page, executed synchronously and fast. This does not cause FOUC.

Thanks Would you suggest that this solution is better than using `<link rel="component.css">` ?

**Comment by HaniTaher** [comment]
Any update on this? We are encouting a big chunk of html, rendred SSR with a lot of styles.
